c7n/resources/asg.py

```diff
@@ -689,6 +690,74 @@ def process(self, asgs, event=None):
         return super(VpcIdFilter, self).process(asgs)
 
 
+@filters.register('launch-activity')
+class LaunchActivityFilter(Filter):
+    """Filter ASG based on launch activity status
+
+    :example:
+
+        .. code-block: yaml
+
+            policies:
+              - name: asg-launch-failures-weekly
+                resource: asg
+                filters:
+                  - type: launch-activity
+                    status: ['Failed']
+                    days: 7
+    """
+    schema = type_schema(
+        'launch-activity',
+        status={
+            'type': 'array',
+            'items': {
+                'enum': [
+                    'Successful', 'Failed', 'Cancelled', 'PreInService',
+                    'InProgress', 'PendingSpotBidPlacement',
+                    'WaitingForSpotInstanceRequestId', 'WaitingForInstanceId',
+                    'WaitingForSpotInstanceId', 'MidLifecycleAction',
+                    'WaitingForELBConnectionDraining',
+                    'WaitingForInstanceWarmup']}},
+        days={'type': 'number'},
+        required=('status',))
+    permissions = ('autoscaling:DescribeScalingActivities',)
+
+    @worker
+    def asg_activities(self, asg):
+        client = local_session(
+            self.manager.session_factory).client('autoscaling')
+        p = client.get_paginator('describe_scaling_activities')
+        expire = False
+        for activity in p.paginate(
+                AutoScalingGroupName=asg['AutoScalingGroupName']):
+            for a in activity['Activities']:
+                if self.data.get('days'):
+                    # if value for days is provided, break after reaching an
+                    # event that is older that desired
+                    if a['StartTime'].replace(tzinfo=tzutc()) < datetime.now(
+                            tz=tzutc()) - timedelta(days=self.data.get('days')):
+                        expire = True
+                        break
+
+                asg.setdefault('c7n:Status', [])
+                if str(a['StatusCode']) not in asg['c7n:Status']:
```

Comment:
why str() on a str()? also your basically emulating .a set... might be simpler to just use a set and convert back to list for json serialize when your done.